fix(service-worker): allow checking for updates when constantly polling the server

gkalpak opened the pull request:
Previously, the SW would wait to become idle before executing scheduled tasks (including checks for newer app versions). It was considered idle when it hadn't received any request for at least 5 seconds. As a result, if the app performed polling (i.e. sent requests to the server) in a shorter than 5 seconds interval, the SW would never detect and update to a newer app version.  
Related issue: #40207

This PR fixes this by adding a max delay to `IdleScheduler` to ensure that no scheduled task will remain pending for longer than the specified max delay.

##
This PR also includes some minor refactorings/improvements. See the individual commits for more details.

mary-poppins commented:
You can preview 6812895 at https://pr40234-6812895.ngbuilds.io/.

mary-poppins commented:
You can preview 5954d49 at https://pr40234-5954d49.ngbuilds.io/.

sonukapoor reviewed (approved):
LGTM 👍

IgorMinar commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -722,18 +718,16 @@ export class Driver implements Debuggable, UpdateSource {
+        cacheNames
+            // The Chrome debugger is not able to render the syntax properly when the
+            // code contains backticks. This is a known issue in Chrome and they have an
+            // open [issue](https://bugs.chromium.org/p/chromium/issues/detail?id=659515) for that.
```
nit: this issue is now marked as fixed

gkalpak replied:
Nice! I removed the work-around :tada:

IgorMinar reviewed (approved):
LGTM!

gkalpak committed:
refactor(service-worker): avoid unnecessarily creating Promises in `Driver#deleteAllCaches()`

This commit refactors `Driver#deleteAllCaches()` to use `Array#map()`
instead of `Array#reduce()` for running async operations in parallel.
This allows avoiding having to recursively wrap Promises with
`Promise.all()`.

gkalpak committed:
fix(service-worker): ensure SW stays alive while notifying clients about unrecoverable state

Previously, the `Driver#notifyClientsAboutUnrecoverableState()` method
would not wait for the completion of the promises created to notify the
clients. Theoretically, this could result in the SW instance's getting
destroyed by the browser before all clients have been notified. This is
extremely unlikely to happen in practice, since the async operations are
very quick, but it _is_ theoretically possible.

This commit ensures that the SW instance will remain alive while
notifying the clients by making `notifyClientsAboutUnrecoverableState()`
await the notification promises.

gkalpak committed:
refactor(service-worker): notify clients about updates in parallel

Previously, clients were notified about updates sequentially. This
wasn't necessary.

This commit changes the `Driver#notifyClientsAboutUpdate()` method to
notify the clients in parallel (by switching from `Array#reduce()` to
`Array#map()` and `Promise.all()`).

This also aligns the `notifyClientsAboutUpdate()` method with the
`notifyClientsAboutUnrecoverableState()` method.

gkalpak committed:
fix(service-worker): allow checking for updates when constantly polling the server

Previously, the SW would wait to become idle before executing scheduled
tasks (including checks for newer app versions). It was considered idle
when it hadn't received any request for at least 5 seconds. As a result,
if the app performed polling (i.e. sent requests to the server) in a
shorter than 5 seconds interval, the SW would never detect and update to
a newer app version.
Related issue: #40207

This commit fixes this by adding a max delay to `IdleScheduler` to
ensure that no scheduled task will remain pending for longer than the
specified max delay.

gkalpak committed:
refactor(service-worker): simplify `Driver#handleFetch()` method

This commit refactors the `Driver#handleFetch()` method to not have to
call `event.waitUntil(this.idle.trigger())` in multiple places.

gkalpak committed:
refactor(service-worker): remove work-around for Chrome code highlighting bug

Chrome debugger code highlighting bug [659515][1] has been fixed, so we
can remove the work-around.
(See #38332 for more details on the work-around.)

[1]: https://bugs.chromium.org/p/chromium/issues/detail?id=659515

mary-poppins commented:
You can preview eaa7cf2 at https://pr40234-eaa7cf2.ngbuilds.io/.
